### ISSUE TITLE
RuntimeException not able to deal with type Light

### ISSUE THREAD

**Issue by jonathanmetzman** [google/graphicsfuzz#274]
This an issue that's come up when I've tried to mutate arbitrary shaders with GraphicsFuzz:

Download [report.zip](https://github.com/google/graphicsfuzz/files/2886650/report.zip) then do:
```
$ mkdir /tmp/report && unzip report.zip -d /tmp/report && mkdir -p /tmp/output
$ java -ea -cp ~/graphicsfuzz-build/jar/tool-1.0.jar com.graphicsfuzz.generator.tool.GlslGenerate /tmp/report ~/graphicsfuzz/shaders/src/main/glsl/samples/donors/donors 1 a_shader_family /tmp/output
13:20:11.663 INFO  GlslGenerate - About to generate 1 shader family, each with 1 variant.
13:20:11.664 INFO  GlslGenerate - Generating family 0 from reference 4af73dcae491da0b94a37e2b6ae4eb54-0.json.
13:20:11.688 INFO  ExecHelper ~/graphicsfuzz-build/bin/Linux/glslangValidator -E --stdin -S frag
13:20:11.700 INFO  ExecHelper - Result: 0
java.lang.RuntimeException: Not able to deal with type Light
        at com.graphicsfuzz.common.util.PipelineInfo.setUniforms(PipelineInfo.java:125)
        at com.graphicsfuzz.common.util.PipelineInfo.zeroUnsetUniforms(PipelineInfo.java:170)
        at com.graphicsfuzz.generator.tool.PrepareReference.prepareReferenceShader(PrepareReference.java:149)
        at com.graphicsfuzz.generator.tool.PrepareReference.prepareReference(PrepareReference.java:131)
        at com.graphicsfuzz.generator.tool.PrepareReference.prepareReference(PrepareReference.java:116)
        at com.graphicsfuzz.generator.tool.GenerateShaderFamily.mainHelper(GenerateShaderFamily.java:160)
        at com.graphicsfuzz.generator.tool.GlslGenerate.mainHelper(GlslGenerate.java:116)
        at com.graphicsfuzz.generator.tool.GlslGenerate.main(GlslGenerate.java:168)
```

This doesn't look like an unknown issue though, sorry if it isn't useful.

**Comment by jonathanmetzman** [comment]
I realize the above may be my own fault since Light is a uniform but I don't specify values for it in the json file.
If I make Light a non-uniform ([attached](https://github.com/google/graphicsfuzz/files/2886660/non-uniform.zip)) I get a new problem:

```
java.lang.AssertionError
        at com.graphicsfuzz.generator.transformation.DonateCodeTransformation.<init>(DonateCodeTransformation.java:100)
        at com.graphicsfuzz.generator.transformation.DonateDeadCodeTransformation.<init>(DonateDeadCodeTransformation.java:62)
        at com.graphicsfuzz.generator.tool.Generate.populateTransformations(Generate.java:508)
        at com.graphicsfuzz.generator.tool.Generate.applyTransformationsMultiPass(Generate.java:389)
        at com.graphicsfuzz.generator.tool.Generate.transformShader(Generate.java:290)
        at com.graphicsfuzz.generator.tool.Generate.generateVariant(Generate.java:184)
        at com.graphicsfuzz.generator.tool.Generate.generateVariant(Generate.java:232)
        at com.graphicsfuzz.generator.tool.GenerateShaderFamily.mainHelper(GenerateShaderFamily.java:228)
        at com.graphicsfuzz.generator.tool.GlslGenerate.mainHelper(GlslGenerate.java:116)
        at com.graphicsfuzz.generator.tool.GlslGenerate.main(GlslGenerate.java:168)
```

**Comment by paulthomson** [comment]
Related #241

**Comment by jonathanmetzman** [comment]
@paulthomson Looks like you were correct about there being a bug in my reproduction instructions I posted.
I still get the problem I first reported ("Not able to deal with") when I run:
```
$ java -ea -cp graphicsfuzz-build/jar/tool-1.0.jar com.graphicsfuzz.generator.tool.GlslGenerate /tmp/report graphicsfuzz-build/shaders/samples/donors/ 1 a_shader_family /tmp/output
...
java.lang.RuntimeException: Not able to deal with type Light
        at com.graphicsfuzz.common.util.PipelineInfo.setUniforms(PipelineInfo.java:125)
        at com.graphicsfuzz.common.util.PipelineInfo.zeroUnsetUniforms(PipelineInfo.java:170)
        at com.graphicsfuzz.generator.tool.PrepareReference.prepareReferenceShader(PrepareReference.java:149)
        at com.graphicsfuzz.generator.tool.PrepareReference.prepareReference(PrepareReference.java:131)
        at com.graphicsfuzz.generator.tool.PrepareReference.prepareReference(PrepareReference.java:116)
        at com.graphicsfuzz.generator.tool.GenerateShaderFamily.mainHelper(GenerateShaderFamily.java:160)
        at com.graphicsfuzz.generator.tool.GlslGenerate.mainHelper(GlslGenerate.java:116)
        at com.graphicsfuzz.generator.tool.GlslGenerate.main(GlslGenerate.java:168)
```

**Comment by paulthomson** [comment]
Thanks for checking! Regardless, we have two bugs here that we should fix.

**Comment by afd** [comment]
Fixed via #419 (at least enough that the struct uniform is just ignored)
